spec.html

```diff
@@ -17,6 +17,107 @@ <h1>Introduction</h1>
     <p>Top-Level Await allows the `await` keyword to be used at the top level of the module goal. See <a href="https://github.com/tc39/proposal-top-level-await/blob/HEAD/README.md">the explainer</a> for the motivation, context, and high-level semantics.</p>
 </emu-intro>
 
+<emu-clause id="sec-code-realms">
+  <h1>Realms</h1>
+  <table>
+    <tbody>
+    <tr>
+      <th>
+        Field Name
+      </th>
+      <th>
+        Value
+      </th>
+      <th>
+        Meaning
+      </th>
+    </tr>
+    <tr>
+      <td>
+        [[Intrinsics]]
+      </td>
+      <td>
+        Record whose field names are intrinsic keys and whose values are objects
+      </td>
+      <td>
+        The intrinsic values used by code associated with this realm
+      </td>
+    </tr>
+    <tr>
+      <td>
+        [[GlobalObject]]
+      </td>
+      <td>
+        Object
+      </td>
+      <td>
+        The global object for this realm
+      </td>
+    </tr>
+    <tr>
+      <td>
+        [[GlobalEnv]]
+      </td>
+      <td>
+        Lexical Environment
+      </td>
+      <td>
+        The global environment for this realm
+      </td>
+    </tr>
+    <tr>
+      <td>
+        [[TemplateMap]]
+      </td>
+      <td>
+        A List of Record { [[Site]]: Parse Node, [[Array]]: Object }.
+      </td>
+      <td>
+        <p>Template objects are canonicalized separately for each realm using its Realm Record's [[TemplateMap]]. Each [[Site]] value is a Parse Node that is a |TemplateLiteral|. The associated [[Array]] value is the corresponding template object that is passed to a tag function.</p>
+        <emu-note>Once a Parse Node becomes unreachable, the corresponding [[Array]] is also unreachable, and it would be unobservable if an implementation removed the pair from the [[TemplateMap]] list.</emu-note>
+      </td>
+    </tr>
+    <tr>
+      <td>
+        <ins>[[AsyncModuleExecIndex]]</ins>
```

Comment:
I'm not quite sure what to call this as there is no precendent for global numberings in the spec.

I did quite like the idea of defining it only to be greater than any in-flight execution, but perhaps that is too nuanced and simply treating it as a global module exec count would be more straightforward.

Reply:
Guy, consider using language like "in the order in which they were marked as blocked".

Often the spec is hard to comprehend, and easier to implement without understanding it, because it is written in such a low-level procedural style. This affects promises and modules especially.

There are opportunities to use more declarative language, to make it much clearer what the spec is trying to do, without introducing imprecision or nondeterminism.

To be clear this is just a difference in style—the specification would remain semantically the same. But I think it would have effects. More people would be able to understand the spec—in terms of time investment and in terms of literal mental capacity. I count myself in this category. Implementation techniques may end up differing more, which I consider a net positive. (In this case, for example, an implementation might be able to find a way to keep the modules in a sorted data structure, possibly eliminating the O(n log n) sort.) The only drawback for the standard's purposes is that these tables would no longer necessarily exhaustively cover every bit of state in a Realm. But I doubt that is particularly helpful anyway, once you get this far into details of particular features.

Reply:
@jorendorff thanks for the feedback on this, any way we can improve or clarify are very welcome.

Are you suggesting for this case not to mark it on the realm, and instead define the execIndex on the module records as something that is just a more general monotonic comparison over all in flight module records? Or just that we change the description of this field here to capture that the ordering is the post-order of the DFS evaluation?

Reply:
I've also added more notes to some of the intermediate functions explaining their roles.

Reply:
I'm proposing dropping Realm.[[AsyncModuleExecIndex]] and Module.[[AsyncExecIndex]]. I believe they are both used solely to specify the ordering in AsyncModuleExecutionFulfilled. But integers are not the only way to describe an ordering. I think the same ordering can be specified unambiguously in prose, and it would be clearer.

Instead of these steps in InnerModuleEvaluation:

> 1. If _module_.[[PendingAsyncDependencies]] &gt; 0 or _module_.[[Async]] is *true*, then
>   1. Let _execIndex_ be _module_.[[Realm]].[[AsyncModuleExecIndex]].
>   1. Set _module_.[[AsyncExecIndex]] to _execIndex_.
>   1. Set _module_.[[Realm]].[[AsyncModuleExecIndex]] to _execIndex_ + 1.
>   1. Set _module_.[[AsyncEvaluating]] to *true*.

I think it could say:

> 1. If _module_.[[PendingAsyncDependencies]] &gt; 0 or _module_.[[Async]] is *true*, then
>   1. Assert: _module_.[[AsyncEvaluating]] is *false* and has never been *true*.
>   1. Set _module_.[[AsyncEvaluating]] to *true*. NOTE: The module becomes blocked. The order in which modules become blocked is significant; see AsyncModuleExecutionFulfilled.

And in AsyncModuleExecutionFulfilled, instead of:

> 1. Let _sortedExecList_ be a List of elements that are the elements of _execList_ ordered by the [[AsyncModuleExecIndex]] field of the Cyclic Module Records in ascending numerical order.

it could say:

> 1. Assert: each element of _execList_ is a Cyclic Module Record with the [[AsyncEvaluating]] field set to *true*.
> 1. Let _sortedExecList_ be a List of the elements of _execList_ , in the order in which their [[AsyncEvaluating]] fields were set to *true*.

If this doesn't seem obviously better, no need to bother with it! But to me this is clearer, and the information is spread across two places in the spec rather than four.

Reply:
This is much nicer yes, I wouldn't have thought that would be have been a well-defined enough definition, but it is certainly a precise one and if it works then certainly let's do this. Thanks for the suggestion!